tests/chainer_tests/test_variable.py

```diff
@@ -1420,6 +1420,84 @@ def test_bool_inconvertible(self):
                 pass
 
 
+_set_data_backend_params = [
+    # NumPy
+    {},
+    # ideep
+    {'use_ideep': 'always'},
+    # CuPy
+    {'use_cuda': True, 'cuda_device': 0},
+    {'use_cuda': True, 'cuda_device': 1},
+    # ChainerX
+    {'use_chainerx': True, 'chainerx_device': 'native:0'},
+    {'use_chainerx': True, 'chainerx_device': 'cuda:0'},
+    {'use_chainerx': True, 'chainerx_device': 'cuda:1'},
+]
+
+
+@testing.parameterize(*testing.product(
+    {
+        'x_shape': [(10,), ()],  # TODO(okapies): add tests for None
+        'requires_grads': [
+            (False, False), (True, False), (False, True), (True, True),
+        ],
+    }))
+@testing.backend.inject_backend_tests(None, _set_data_backend_params)
+@testing.backend.inject_backend_tests(None, _set_data_backend_params)
+class TestVariableSetArray(unittest.TestCase):
+
+    def _gen_array(self, device, shape, requires_grad):
+        if shape is None:
+            return None
+
+        a = device.send_array(
+            np.random.uniform(-1, 1, shape).astype(np.float32))
+        assert device.is_array_supported(a)
+
+        if device.xp is chainerx and requires_grad:
+            return a.require_grad()
+        else:
+            return a
+
+    def test_set_array(self, backend_config1, backend_config2):
+        device_x = backend_config1.device
+        device_y = backend_config2.device
+        xp_x = device_x.xp
+        requires_grad_x = self.requires_grads[0]
+        requires_grad_y = self.requires_grads[1]
+
+        x = self._gen_array(device_x, self.x_shape, requires_grad_x)
+        y = self._gen_array(device_y, self.x_shape, requires_grad_y)
+
+        v = chainer.Variable(x, requires_grad=requires_grad_x)
+
+        if isinstance(device_x, backend.GpuDevice):
+            should_fail = device_x != device_y
+        elif isinstance(device_x, backend.ChainerxDevice):
+            should_fail = (
+                device_x != device_y or requires_grad_x or requires_grad_y)
+        elif (isinstance(device_x, backend.Intel64Device)
+              or isinstance(device_y, backend.Intel64Device)):
+            # Check the array type because Intel64Device.send_array() may
+            # return numpy.ndarray when its shape doesn't match
+            if isinstance(x, intel64.mdarray):
+                should_fail = not isinstance(y, (np.ndarray, intel64.mdarray))
+            else:
+                should_fail = type(x) != type(y)
```

Comment:
`(type(x), type(y)) == (numpy.ndarray, intel64.mdarray)`
Shouldn't this case succeed?